client/src/JobPaginatorState.js

```diff
@@ -0,0 +1,305 @@
+    if (order.length < 2) {
+        alert("You cannot delete the last page in a job.");
+        return false;
+    }
+    return window.confirm("Are you sure you wish to delete the current page?");
+};
+
+const getAddedPages = (order, pages) => {
+    return pages.reduce((order, page) => {
+        order.splice(findNewPagePosition(page, order), 0, {
+            filename: page,
+            label: null,
+        });
+        return order;
+    }, order);
+};
+
+const getNoneRemovedPages = (order, deletedPages) =>
+    order.filter((o) =>
+        deletedPages.every((page) => {
+            return o.filename !== page;
+        })
+    );
+
+const validatePublish = async (category, job) => {
+    const response = await getStatus(category, job);
+    const {
+        derivatives: { expected, processed },
+    } = response;
+    if (expected > processed) {
+        alert(
+            `Derivative images have not been generated yet. Please` +
+                ` go back to the main menu and hit the "build" button` +
+                ` for this job before publishing it.`
+        );
+        return false;
+    }
+    return !window.confirm(
+        "Are you sure you wish to publish this job? You will not be able to make any further edits."
+    );
+};
+
+const getLabel = (order, imageNumber) => {
+    return typeof order[imageNumber] === "undefined" ? null : order[imageNumber]["label"];
+};
+
+const countMagicLabels = (order, startAt) => {
+    let count = 0;
+    for (let startIndex = startAt; startIndex < order.length; startIndex++) {
+        if (!getLabel(order, startIndex)) {
+            count++;
+        }
+    }
+    return count;
+};
+
+const validateSave = (order) => {
+    const count = countMagicLabels(order, 0);
+    return count > 0 && !confirmSavedMagicLabels(count);
+};
+
+const useJobPaginator = (initialCategory, initialJob) => {
+    const [active, setActive] = useState(false);
+    const [currentPage, setCurrentPage] = useState(0);
+    const [zoom, setZoom] = useState(false);
+    const [order, setOrder] = useState([]);
+    const [magicLabelCache, setMagicLabelCache] = useState([]);
+    const [category, setCategory] = useState(initialCategory);
+    const [job, setJob] = useState(initialJob);
+
+    const saveMagicLabels = () => {
+        order.forEach((o, orderIndex) => {
+            if (!getLabel(order, orderIndex)) {
+                setLabel(orderIndex, getMagicLabel(orderIndex));
+            }
+        });
+    };
+
+    const setLabel = (imageNumber, text) => {
+        setMagicLabelCache([]); // clear label cache whenever there is a change
+        if (getLabel(order, imageNumber)) {
+            return;
+        }
+        order[imageNumber]["label"] = text !== null && text.length === 0 ? null : text;
+        setOrder(order);
+        dispatchEvent(new Event("Prep.editted"));
+    };
+
+    const getMagicLabel = (imageNumber) => {
+        const label = getLabel(order, imageNumber);
+
+        if (!label) {
```

Comment:
Is this logic backwards? It seems like it will cause this to always return null (or at least return null under inappropriate circumstances).

I appreciate your efforts to make the getLabel() method less confusing and more explicit, but I think some subtlety has gotten lost along the way. There are really three scenarios that need to be addressed:

1.) We want to know whether or not the user has set a label explicitly on an image.
2.) We want to get the auto-generated "magic label" for a set position.
3.) We want to get the user-generated label if available, or the "magic label" otherwise.

I think in the current code, cases 2 and 3 are getting mixed up and it's causing problems. Maybe we want three functions: getUserLabel, getMagicLabel, and getBestLabel... that might make this all more clear.

Reply:
Yes. I realized that after applying the unit tests. I reserved the logic